tests/federation/transport/test_knocking.py

```diff
@@ -0,0 +1,289 @@
+from collections import OrderedDict
+from typing import Dict, List
+
+from mock import Mock
+
+from twisted.internet.defer import succeed
+
+from synapse import event_auth
+from synapse.api.constants import EventTypes
+from synapse.api.room_versions import RoomVersions
+from synapse.config.ratelimiting import FederationRateLimitConfig
+from synapse.events import builder
+from synapse.federation.transport import server as federation_server
+from synapse.rest import admin
+from synapse.rest.client.v1 import login, room
+from synapse.server import HomeServer
+from synapse.types import RoomAlias
+from synapse.util.ratelimitutils import FederationRateLimiter
+
+from tests.test_utils import event_injection, make_awaitable
+from tests.unittest import FederatingHomeserverTestCase, HomeserverTestCase
+
+# An identifier to use while MSC2304 is not in a stable release of the spec
+KNOCK_UNSTABLE_IDENTIFIER = "xyz.amorgan.knock"
```

Comment:
I think we already have this as a constant somewhere?

Reply:
We have it as a join rule and membership event in `synapse.api.constants`, however those don't exactly line up with endpoint paths. I realise I've been using this for join rules and membership types though, so those will be replaced!